opt_einsum/paths.py

```diff
@@ -841,53 +915,57 @@ def __call__(self, inputs, output, size_dict, memory_limit=None):
         >>>             i[k].add(c)
         >>>             s[c] = 2
         >>>     i_all.extend(i)
-        >>> o = DynamicProgrammingOptimizer()
+        >>> o = DynamicProgramming()
         >>> o(i_all, set(), s)
         [(1, 2), (0, 4), (1, 2), (0, 2), (0, 1)]
         """
+        ind_counts = Counter(itertools.chain(*inputs, output))
+        all_inds = tuple(ind_counts)
 
         # convert all indices to integers (makes set operations ~10 % faster)
-        symbol2int = {c: j for j, c in enumerate(set.union(*inputs) | output)}
+        symbol2int = {c: j for j, c in enumerate(all_inds)}
         inputs = [set(symbol2int[c] for c in i) for i in inputs]
         output = set(symbol2int[c] for c in output)
         size_dict = {symbol2int[c]: v for c, v in size_dict.items() if c in symbol2int}
         size_dict = [size_dict[j] for j in range(len(size_dict))]
 
-        # all summation indices occurring exactly in one input:
-        i_single = set(
-            c for c in set.union(*inputs) - output
-            if sum(1 for i in inputs if c in i) == 1
-        )
-
-        # contraction expressions for all inputs that have already been
-        # reduced to scalars:
-        inputs_done = [
-            (j,) for j, i in enumerate(inputs)
-            if len(i - i_single) == 0
-        ]
-
-        # remaining input index sets and corresponding contraction expressions;
-        # indices from i_single are removed and if a single-tensor contraction
-        # is performed, the contraction expression is (j,) instead of j;
-        inputs, inputs_contractions = zip(*[
-            (i - i_single, j if i.isdisjoint(i_single) else (j,))
-            for j, i in enumerate(inputs)
-            if len(i - i_single) > 0
-        ])
+        # parse all summation indices occurring exactly in one input
+        i_single = {i for i, c in enumerate(all_inds) if ind_counts[c] == 1}
+        enum_inputs = enumerate(inputs)
+        inputs, inputs_done, inputs_contractions = [], [], []
+        for j, i in enum_inputs:
+            i_reduced = i - i_single
+            if not i_reduced:
+                # input reduced to scalar already - remove
+                inputs_done.append((j,))
+            else:
+                # if the input has any index reductions, add single contraction
+                inputs.append(i_reduced)
+                inputs_contractions.append((j,) if i_reduced != i else j)
```

Comment:
I see that my original code did not work if there was nothing more to optimise at this point. However, I find this reassignment of `inputs` in combination with the loop rather confusing.. I guess the original code could be fixed, if  the `zip` would only be unpacked if the list constructed inside the `zip` is not empty.

Reply:
Yes this seemed liked the simplest way to solve the 'nothing left to do' bug and also not iterate and filter the inputs twice in the same way. But certainly open to suggestions for making it more readable. Maybe just renaming the new ``inputs ->  inputs_parsed`` or something?

Reply:
To be honest, for me list comprehensions are much more expressive compared to list constructions using loops. However, at least the variable `inputs` should be renamed.

Reply:
Of course generally I agree! But when for loops can do the same thing more efficiently... I think they are the correct tool to use. I've renamed the parsed ``inputs`` to ``inputs_contract``.